lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp

```diff
@@ -114,6 +114,68 @@ struct EnableGroupingPattern : public OpRewritePattern<ClockTreeOp> {
     return success(changed);
   }
 };
+
+struct GroupAssignmentsInIfPattern : public OpRewritePattern<ClockTreeOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(ClockTreeOp clockTreeOp,
+                                PatternRewriter &rewriter) const override {
+    // Pull values only used in certain reset/enable cases into the appropriate
+    // IfOps
+    SmallVector<Region *> groupingRegions;
+    // Gather then/else regions of all top-level IfOps
+    for (auto resetIfOp : clockTreeOp.getBody().getOps<scf::IfOp>()) {
+      groupingRegions.push_back(&resetIfOp.getThenRegion());
+      groupingRegions.push_back(&resetIfOp.getElseRegion());
+    }
+    // Gather then/else regions of all second-level IfOps (e.g. enables within
+    // resets)
+    SmallVector<Region *> secondLevelRegions;
+    for (auto *resetRegion : groupingRegions) {
+      for (auto enableIfOp : resetRegion->getOps<scf::IfOp>()) {
+        secondLevelRegions.push_back(&enableIfOp.getThenRegion());
+        secondLevelRegions.push_back(&enableIfOp.getElseRegion());
+      }
+    }
+    groupingRegions.insert(groupingRegions.end(), secondLevelRegions.begin(),
+                           secondLevelRegions.end());
+
+    bool changed = false;
+    for (auto *region : groupingRegions) {
+      if (region->empty())
+        continue;
+      // Since we only work with IfOp then/else regions, we have at most 1 block
+      Block &block = region->front();
+      for (auto &op : block) {
+        for (auto operand : op.getOperands()) {
+          // If the op is already in the region or the definition is outside the
+          // clock tree, skip
+          Operation *definition = operand.getDefiningOp();
+          if (definition->getParentRegion() == region ||
+              !clockTreeOp->isAncestor(definition))
+            continue;
+          // If the operand is used more than once, check moving the assignment
+          // wouldn't cause non-domination
+          if (!operand.hasOneUse()) {
+            bool safeToMove = true;
+            for (auto *user : operand.getUsers()) {
+              if (!region->isAncestor(user->getParentRegion())) {
+                safeToMove = false;
+                break;
+              }
+            }
+            if (!safeToMove)
+              break;
+          }
+          rewriter.updateRootInPlace(definition,
+                                     [&]() { definition->moveBefore(&op); });
+          changed = true;
+        }
```

Comment:
I think it might be better to do a DFS/BFS traversal of the SSA use/def chain on each operand defined outside the current region (but inside the clock_tree) until the first op that also has uses outside the current region is found. This way many operations can be pulled in per pattern match which speeds things up when a lot of ops can be pulled in and the implementation isn't a lot more difficult, just a worklist and a while loop.

Reply:
This is a great point, I'll give that a try, definitely seems more efficient!